### ISSUE TITLE
API versioning strict mode not returning a 406 as expected

### ISSUE THREAD

**Issue by nicholaspufal** [ruby-grape/grape#1413]
I created a very simple endpoint to make it easier to spot this issue. Say I have the following API:

```ruby
module MyApp
  class API < Grape::API
    format :json
    version 'v1', using: :accept_version_header, strict: true

    get 'healthcheck' do
      status 200
    end
  end
end
```

And the following tests:

```ruby
describe 'MyApp::API' do
  def app
    MyApp::API
  end

  describe 'versioning' do
    it 'fails if the client does not specify a version' do
      get '/healthcheck'

      expect(last_response.status).to eq(406)
    end

    it 'returns successfully if a version if specified' do
      header 'Accept-Version', 'v1'

      get '/healthcheck'

      expect(last_response.status).to eq(200)
    end
  end
end
```

What's happening is that for the first scenario it's returning a `404`. 

After investigating a bit deeper the issue it seems like that `versioner/accept_version_header.rb` is doing the right job - I can see a 406 from there, however later in the process `Grape::Middleware::Error#error_response` appends a `X-Cascade: pass` to the headers and that seems to be used later by `Grape::Router#transaction` to search for another match of that route (lines 86 to 90 in `lib/grape/router.rb`). As the router can't find it ends up returning a `404`.

I'm not exactly sure what's suppose to be going on as part of that last stage, so maybe someone with more context on it can help me here.

The workaround for now is commented out as part of https://github.com/ruby-grape/grape/blob/master/lib/grape/middleware/versioner/accept_version_header.rb#L48

So in my example, it would mean that this line:

```ruby
version 'v1', using: :accept_version_header, strict: true
```

...would become this:

```ruby
version 'v1', using: :accept_version_header, strict: true, cascade: false
```

That work wonders but it's still unclear to me why that behavior is not default.

Thoughts?

Thanks!
Nicholas

**Comment by nicholaspufal** [comment]
After taking another look at the docs, this behavior is actually being mentioned:

>When an invalid Accept header is supplied, a 406 Not Acceptable error is returned if the :cascade option is set to false. Otherwise a 404 Not Found error is returned by Rack if no other route matches.

Wouldn't the other way around make more sense? Make the default `cascade: false`?

**Comment by dblock** [comment]
It could make more sense, but I bet the reason is historical and I would vote not to change this behavior so that everybody doesn't have to suffer. If you agree I think we should close this?

**Comment by nicholaspufal** [comment]
Yeah, I thought about the historical reasons as well. 

Before closing it, could I make a small change to the docs? 

I'm thinking about copying these lines from the **Header** explanation:

> When an invalid Accept header is supplied, a 406 Not Acceptable error is returned if the :cascade option is set to false. Otherwise a 404 Not Found error is returned by Rack if no other route matches.

And use them as part of the **Accept-Version Header** docs - at the moment it only mentions that when using `strict: true` it will return a `406`.

**Comment by dblock** [comment]
Definitely propose a PR, thanks @nicholaspufal!

**Comment by dblock** [comment]
Closed via #1417.
